src/field_theory/algebraic_closure.lean

```diff
@@ -0,0 +1,283 @@
+  ∃ x : adjoin_monic k, f.eval₂ (to_adjoin_monic k) x = 0 :=
+⟨ideal.quotient.mk _ $ X (⟨f, hfm, hfi⟩ : monic_irreducible k),
+ by { rw [to_adjoin_monic, ← hom_eval₂, ideal.quotient.eq_zero_iff_mem],
+  exact le_max_ideal k (ideal.subset_span $ ⟨_, rfl⟩) }⟩
+
+/-- The auxiliary function to the `n`th step. -/
+def step_aux (n : ℕ) : Σ α : Type u, field α :=
+nat.rec_on n ⟨k, infer_instance⟩ $ λ n ih, ⟨@adjoin_monic ih.1 ih.2, @adjoin_monic.field ih.1 ih.2⟩
+
+/-- The `n`th step. -/
+def step (n : ℕ) : Type u :=
+(step_aux k n).1
+
+instance step.field (n : ℕ) : field (step k n) :=
+(step_aux k n).2
+
+instance step.inhabited (n) : inhabited (step k n) := ⟨37⟩
+
+/-- The canonical inclusion to the `0`th step. -/
+def to_step_zero : k →+* step k 0 :=
+ring_hom.id k
+
+/-- The canonical ring homomorphism to the next step. -/
+def to_step_succ (n : ℕ) : step k n →+* step k (n + 1) :=
+@to_adjoin_monic (step k n) (step.field k n)
+
+instance step.algebra_succ (n) : algebra (step k n) (step k (n + 1)) :=
+(to_step_succ k n).to_algebra
+
+theorem to_step_succ.exists_root {n} {f : polynomial (step k n)}
+  (hfm : f.monic) (hfi : irreducible f) :
+  ∃ x : step k (n + 1), f.eval₂ (to_step_succ k n) x = 0 :=
+@adjoin_monic.exists_root _ (step.field k n) _ hfm hfi
+
+/-- The canonical ring homomorphism to a step with a greater index. -/
+def to_step_of_le (m n : ℕ) (h : m ≤ n) : step k m →+* step k n :=
+{ to_fun := nat.le_rec_on h (λ n, to_step_succ k n),
+  map_one' := begin
+    induction h with n h ih, { exact nat.le_rec_on_self 1 },
+    rw [nat.le_rec_on_succ h, ih, ring_hom.map_one]
+  end,
+  map_mul' := λ x y, begin
+    induction h with n h ih, { simp_rw nat.le_rec_on_self },
+    simp_rw [nat.le_rec_on_succ h, ih, ring_hom.map_mul]
+  end,
+  map_zero' := begin
+    induction h with n h ih, { exact nat.le_rec_on_self 0 },
+    rw [nat.le_rec_on_succ h, ih, ring_hom.map_zero]
+  end,
+  map_add' := λ x y, begin
+    induction h with n h ih, { simp_rw nat.le_rec_on_self },
+    simp_rw [nat.le_rec_on_succ h, ih, ring_hom.map_add]
+  end }
+
+@[simp] lemma coe_to_step_of_le (m n : ℕ) (h : m ≤ n) :
+  (to_step_of_le k m n h : step k m → step k n) = nat.le_rec_on h (λ n, to_step_succ k n) :=
+rfl
+
+instance step.algebra (n) : algebra k (step k n) :=
+(to_step_of_le k 0 n n.zero_le).to_algebra
+
+instance step.algebra_tower (n) : is_algebra_tower k (step k n) (step k (n + 1)) :=
+is_algebra_tower.of_algebra_map_eq $ λ z,
+@nat.le_rec_on_succ (step k) 0 n n.zero_le (n + 1).zero_le (λ n, to_step_succ k n) z
+
+theorem step.is_integral (n) : ∀ z : step k n, is_integral k z :=
+nat.rec_on n (λ z, is_integral_algebra_map) $ λ n ih z,
+is_integral_trans ih _ (adjoin_monic.is_integral (step k n) z : _)
+
+instance to_step_of_le.directed_system :
+  directed_system (step k) (λ i j h, to_step_of_le k i j h) :=
+⟨λ i x h, nat.le_rec_on_self x, λ i₁ i₂ i₃ h₁₂ h₂₃ x, (nat.le_rec_on_trans h₁₂ h₂₃ x).symm⟩
+
+end algebraic_closure
+
+/-- Algebraic closure. -/
```

Comment:
```suggestion
/-- The canonical algebraic closure of a field, the direct limit of adding roots to the field for each polynomial over the field. -/
```

Reply:
Is it canonical?

Reply:
I would like the docstring to distinguish this from e.g. the algebraic closure of ℚ as a subfield of ℂ. Perhaps 'canonical' is not the canonical choice of words here.

Reply:
(By `subfield` I mean something of the form `{x : ℂ // ...}`